### DISCUSSION TITLE
Trajectory planning

### DISCUSSION THREAD

**Discussion by galleon** [do-mpc/do-mpc#156] [Q&A]
Hello,

First thanks for the great package that I discovered few days ago!

Is is possible to use it to do trajectory planning? 

I have a problem that has time varying physical quantities in which a system is evolving. First I decided to play with a simpler bicycle model.

I have it located at a certain position at time 0 and want it to reach and different position (at an unknown time t). Apparently it is solved instantly which I don't understand.

I am a bit lost at the moment ! Everything is [here](https://gist.github.com/galleon/f3b496e24257d8aa8ce94ea69f3f2e7e) in cas that can help

Cheers

**Comment by 4flixt** [comment]
Hi @galleon ,

thank you for the compliment and also for making it easy to follow your problem :)
I have to admit that I didn't catch any error immediately. The only problem I saw is that your simulator and your controller use a different time-step. That is generally supported but than you have to simulate multiple steps for each control step in the main loop:
```python
for k in range(10):
  u0 = mpc.make_step(state_0)
  for j in range(mpc.dt//simulator.dt)
     x0 = simulator.make_step(u0)
```
I think the next steps in debugging would be: 
a) Look at the output of a single call of ```mpc.make_step``` - did the optimizer find an optimial solution or is the problem e.g. infeasible? Just call ```mpc.make_step``` in a cell without ``%%capture``.
b) plot the prediction of a single call of ```mpc.make_step```. You can obtain it with: ```mpc.data.prediction(('_x', 'x'))`` for the first state. Is the controller predicting a proper trajectory? If yes then the problem lies in the close-loop application (e.g. because if different time-steps).

I hope this helps! Best 4flixt

**Reply by 4flixt** [reply]
Actually, now that I am looking at my answer I saw the error in your code :D 

Look at your main loop (copy-pasted in my answer). You compute the control input for ``state_0`` and your simulator is returning ``x0``. Thus your loop is not closed.

**Reply by galleon** [reply]
Ouch that was indeed a lame error:
Doing something like this:
```
for k in range(200):
    u0 = mpc.make_step(state_0)
    state_0 = simulator.make_step(u0)
``` 
seems to fix my problem.
What is the way to plot the trajectory - I mean retrieving x & y as as list of points in time ?

Then I need to remove my loop and probably check that I reached my final state. Right ?

Then I will move to my 3d example :-) Thanks for looking into this !

**Comment by 4flixt** [comment]
Hey @galleon ,

have a look [here](https://www.do-mpc.com/en/latest/api/do_mpc.data.MPCData.html#do_mpc.data.MPCData). This explains how to retrieve data from the MPC object. Things are similar for the simulator object.

**Reply by galleon** [reply]
Thanks a ton, that worked.

Glop, after plotting the trajectory it is not at all what I was anticipating. Especially the end seems a bit chaotic ...

![image](https://user-images.githubusercontent.com/259772/117308061-714fc700-ae81-11eb-9fa5-ffe2186d4639.png)

Any idea what might be causing this ?
Any secret recipe ? Is that related to the "shaping" of `lterm` and `mterm` ?

**Reply by 4flixt** [reply]
I think your bike crashed :( 

I fear that at this point you just have to play around a bit with your tuning. It is hard to say what could be improved without investing a lot of time into the specific problem. I hope you'll figure it out and continue using do-mpc :) 

Best 4flixt

**Comment by TheCodeSummoner** [comment]
You can certainly use `do_mpc` to achieve this.

Shaping of `lterm` and `mterm` is definitely important, as it defines what the model will attempt to achieve. I have played around with things like that in my Bachelor's thesis, where I have modelled the bicycle model and asked it to follow target reference points based on Euclidean distance.

You can find the code [here](https://github.com/TheCodeSummoner/f1tenth-racing-algorithms/blob/master/f1tenth/common/mpc.py), perhaps something will be useful for you. I am also sharing a debugging tool I've written in [Python Arcade](https://arcade.academy/) (see below), which uses the Point-Mass model (see [this](https://gitlab.lrz.de/tum-cps/commonroad-vehicle-models/-/blob/master/vehicleModels_commonRoad.pdf) for the definition of the model). I've not included descriptions of each component, nor a version with time-varying parameters, as I believe you should manage to figure things out.

There are also papers that consider for example heading angle fixing (so that your bicycle is always oriented towards your target point) or following a pre-defined reference trajectory (consisting of multiple points).

```python
import arcade
import numpy as np
import do_mpc
import casadi
import matplotlib.pyplot as plt


class Application(arcade.Window):

    def setup_model(self):
        self.model = do_mpc.model.Model("continuous")
        self.position_var = self.model.set_variable(
            var_type="_x",
            var_name="position",
            shape=2
        )
        self.velocity_var = self.model.set_variable(
            var_type="_x",
            var_name="velocity",
            shape=2
        )
        self.acceleration_var = self.model.set_variable(
            var_type="_u",
            var_name="acceleration",
            shape=2
        )
        self.model.set_rhs("position", self.velocity_var)
        self.model.set_rhs("velocity", self.acceleration_var)
        self.model.setup()

    def setup_controller(self):
        self.controller = do_mpc.controller.MPC(self.model)
        self.controller.set_param(
            n_horizon=20,
            t_step=0.2,
            nlpsol_opts={"ipopt.print_level": 0, "ipopt.sb": "yes", "print_time": 0},  # suppress console print
            store_full_solution=True
        )
        lagrange_term = casadi.DM.zeros()  # no stage cost
        mayer_term = (self.target[0] - self.position_var[0]) ** 2 + (self.target[1] - self.position_var[1]) ** 2
        self.controller.set_objective(
            lterm=lagrange_term,
            mterm=mayer_term
        )
        self.controller.set_nl_cons("cons", self.acceleration_var[0] ** 2 + self.acceleration_var[1] ** 2, 100)
        self.controller.setup()
        self.controller.x0 = np.array([self.position[0], self.position[1], self.velocity[0], self.velocity[1]])
        self.controller.set_initial_guess()

    def setup_plotting(self):
        self.graphics = do_mpc.graphics.Graphics(self.controller.data)
        fig, ax = plt.subplots(2, sharex="all", figsize=(16, 9))
        fig.align_ylabels()
        self.graphics.add_line(var_type="_x", var_name="position", axis=ax[0], color="r")
        self.graphics.add_line(var_type="_x", var_name="velocity", axis=ax[1], color="g")
        self.graphics.add_line(var_type="_u", var_name="acceleration", axis=ax[1], color="b")
        ax[0].set_ylabel("Position")
        ax[1].set_ylabel("velocity (green), and acceleration (blue)")
        ax[1].set_xlabel("Time")

    def plot(self):
        self.graphics.plot_results()
        self.graphics.reset_axes()
        plt.show()

    def __init__(self, width, height, title):
        super().__init__(width, height, title)
        arcade.set_background_color(arcade.color.WHITE)
        self.target = np.array([200, 300])
        self.position = np.array([0, 0])
        self.velocity = np.array([0, 0])
        self.acceleration = np.array([1, 1])  # 1, 1 to verify the point is moving
        self.setup_model()
        self.setup_controller()
        self.setup_plotting()

    def on_draw(self):
        arcade.start_render()
        arcade.draw_point(
            x=self.target[0],
            y=self.target[1],
            color=arcade.color.RED,
            size=10
        )
        arcade.draw_point(
            x=self.position[0],
            y=self.position[1],
            color=arcade.color.LIGHT_BLUE,
            size=10
        )

        for point in self.predicted:
            arcade.draw_point(
                x=point[0],
                y=point[1],
                color=arcade.color.LIGHT_BLUE,
                size=4
            )

    def on_update(self, dt: float):
        self.velocity = self.velocity + self.acceleration * dt
        self.position = self.position + self.velocity * dt
        self.acceleration = self.controller.make_step(
            np.array([self.position[0], self.position[1], self.velocity[0], self.velocity[1]])
        ).flatten()
        x_predicted = self.controller.data.prediction(("_x", "position", 0)).flatten()
        y_predicted = self.controller.data.prediction(("_x", "position", 1)).flatten()
        self.predicted = np.dstack((x_predicted, y_predicted))[0]

app = Application(800, 600, "do_mpc example")
arcade.run()
app.plot()
```

**Reply by zmdntnu** [reply]
Hi @TheCodeSummoner,

I hope I get in touch with you through this discussion page. I'm studying your code shown above, and I have a question regarding the on_update function. Is it part of the MPC framework in your case or does it belong to arcade only? And if I want to disable arcade and show only the results from MPC, how can I do that? 


Best regards

**Reply by TheCodeSummoner** [reply]
> Is it part of the MPC framework or does it belong to arcade only?

`on_update` is coming from `arcade.Window`. I recommend using `PyCharm` editor for writing Python code since you could check things like this with a single button click.

> if I want to disable arcade and show only the results from MPC, how can I do that?

What do you mean by showing the results? What are the results anyway? The predicted trajectory? The loss over time? They are all just numbers, so you can display them in whatever form you want to. For example, in the code above, some metrics are also plotted at the end of the program run (`plot` function configured by `setup_plotting`). The easiest is to experiment and gradually understand the framework I'd say.

Keep in mind though that `PyArcade` is giving me an infinite loop which always running the code (`on_update` runs every tick and measures elapsed time). In my thesis, I was using ROS which also had an infinite loop. Generally, you may need something similar in your project, to keep measuring the environment and computing results.
